### PR TITLE
feat: deepseekv4 attn collect and query

### PR DESCRIPTION
**Pull request by YijiaZhao**
Overview:                                                                                                                                                                                                                                                                                                                                        

Add DeepSeek-V4-Flash (sparse attn: CSA/HCA/SWA) performance modeling support to aiconfigurator's sglang backend, including module-level and sparse-kernel data collection, perf_database loaders/queries, and unit tests. The V4-Flash code path is fully isolated — no behavior change for any other model.                                            

  Details:

  Collector (1 OpEntry = 1 file convention)
  - New collector/sglang/collect_dsv4_flash_attn.py: 6 ops (csa/hca/swa × ctx/gen) for full self-attn module timing (projections + attn).
  - New collector/sglang/deepseekv4_sparse_modules.py: 2 sparse-kernel ops (paged_mqa_logits, hca_attn) for kernel-Δ correction.
  - _pick_free_port() via socket.bind(0) replaces random.randint, fixes EADDRINUSE under 8-way concurrent collection.
  - Sparse hca_attn mirrors sglang's deepseek_v4.py:847 zero-pad logic: q_local of shape (b, M, n_local_heads, D) is zero-padded into (b, M, 64, D) so FlashMLA's h_q ∈ {64, 128} constraint is satisfied at any TP. Sweep is SPARSE_TP_LIST_ATTN=[1] since the kernel is TP-invariant after padding.
  - collect.py: when --model-path deepseek-ai/DeepSeek-V4-Flash is given without explicit --ops, auto-expands to all dsv4_flash_* ops and uses short label dsv4_flash for log filename (avoids OSError on long joined names).

  perf_database
  - _dsv4_flash_tp_from_num_heads(num_heads) = 64 // num_heads reverses sglang's pre-pad local head count from collected CSV.
  - _dsv4_flash_robust_3d_lookup: exact → cubic → linear fallback. V4-only; other models keep existing _interp_3d cubic path. Avoids qhull degeneracy on V4's caps-driven flat b*s ≤ 8192 axis.
  - load_context_dsv4_flash_kind_module_data keyed by tp_size; load_generation_dsv4_flash_kind_module_data stores [kv][gemm][arch][cr][tp][b][s_total] to match _interp_3d(num_heads, b, s) arg order.
  - _deep_merge_dsv4_dicts merges 6 split CSA/HCA/SWA ctx+gen files into a single _context_deepseek_v4_attention_module_data attribute, preserving the existing query API.
  - query_context_deepseek_v4_attention_module adds kernel-Δ correction with topk_512 modeled as memory-IO: Δ_bytes = M × past_kv at eff = 0.1.

  Tests — 24 new unit tests in tests/unit/sdk/database/test_dsv4_flash_sparse.py:
  - 5 × _dsv4_flash_tp_from_num_heads reverse derivation
  - 3 × sparse kernel CSV loader (basic / dup-headers / missing)
  - 1 × context loader keying by tp; 1 × generation loader [b][s_total] order
  - 2 × deep_merge; 2 × _robust_3d_lookup (exact + cubic→linear fallback)
  - 4 × sparse kernel lookup (exact / tp fallback / interp / missing)
  - 4 × test-case generators with --model-path filter
  - 2 × topk IO formula

  Verified accuracy (tp8, fp8_block, ISL=8192, h20_sxm)

  ┌────────────┬───────────┬──────────┬───────┐
  │     Op     │ Predicted │   Real   │   Δ   │
  ├────────────┼───────────┼──────────┼───────┤
  │ CSA chunk0 │ 10.55 ms  │ 10.92 ms │ -3.4% │
  ├────────────┼───────────┼──────────┼───────┤
  │ CSA chunk1 │ 11.69 ms  │ 12.06 ms │ -3.0% │
  ├────────────┼───────────┼──────────┼───────┤
  │ HCA chunk0 │ 4.45 ms   │ —        │ -0.2% │
  ├────────────┼───────────┼──────────┼───────┤
  │ HCA chunk1 │ 4.93 ms   │ —        │ -1.0% │
  └────────────┴───────────┴──────────┴───────┘
  Where should the reviewer start?

  1. collector/sglang/deepseekv4_sparse_modules.py — _bench_flash_mla_sparse zero-pad logic (the most subtle piece; mirrors sglang exactly).
  2. src/aiconfigurator/sdk/perf_database.py — _dsv4_flash_robust_3d_lookup, _dsv4_flash_tp_from_num_heads, and query_context_deepseek_v4_attention_module topk-IO + kernel-Δ correction.
  3. collector/sglang/collect_dsv4_flash_attn.py — _pick_free_port + the 6 split get_funcs.
  4. collector/collect.py — auto-expand block (gated on exact model-path string; isolated).
  5. tests/unit/sdk/database/test_dsv4_flash_sparse.py — coverage map for the above.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added comprehensive DeepSeek-V4-Flash benchmarking support including performance measurement for multiple attention types (context and generation modes).
  * Integrated sparse-kernel benchmarking capabilities for specialized operators.
  * Enhanced SDK with V4-Flash performance data integration and intelligent data lookup with fallback strategies.
  * Automatic test case generation for V4-Flash configurations.

* **Tests**
  * Added comprehensive unit tests validating the V4-Flash data pipeline and sparse-kernel functionality.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

File: collector/collect.py [hunk 1]
```diff
@@ -1092,18 +1093,36 @@ def main():
...
+
+        # V4-Flash special-case: when the model is V4-Flash and no explicit
+        # ``--ops`` is given, scope to the V4-Flash ops only (skip generic
+        # gemm / attention / mla / ... that aren't consumed by V4's perf
+        # model).  All other models keep the default behaviour: run every
+        # op and let each get_func's ``_filter_model_config_list`` filter
+        # cases at the test-case level.
+        if not ops and args.model_path == "deepseek-ai/DeepSeek-V4-Flash":
+            ops = [name for name in _all_op_names() if name.startswith("dsv4_flash_")]
+            _dsv4_auto_expand = True
```
[the remark this answers is not in the document]
Comment (YijiaZhao) [reply collector/collect.py]: done

File: collector/sglang/collect_dsv4_flash_attn.py [hunk 2]
```diff
@@ -0,0 +1,1267 @@
...
+ATTN_KIND_TO_COMPRESS_RATIO = {
+    "swa": 0,
+    "csa": 4,
+    "hca": 128,
+}
+
+COMPRESS_RATIO_TO_ATTN_KIND = {v: k for k, v in ATTN_KIND_TO_COMPRESS_RATIO.items()}
+
+DEFAULT_MODEL = "deepseek-ai/DeepSeek-V4-Pro"
+_WEIGHT_SUFFIXES = (".safetensors", ".bin", ".pt", ".pth")
```
[the remark this answers is not in the document]
Comment (YijiaZhao) [reply collector/sglang/collect_dsv4_flash_attn.py]: fixed

File: collector/sglang/deepseekv4_sparse_modules.py [hunk 3]
```diff
@@ -0,0 +1,723 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
```
[the remark this answers is not in the document]
Comment (YijiaZhao) [reply collector/sglang/deepseekv4_sparse_modules.py:2]: fixed

File: collector/sglang/deepseekv4_sparse_modules.py [hunk 4]
```diff
@@ -0,0 +1,723 @@
...
+    from common_test_cases import (
+        _DSV4_FLASH_SPARSE_TP_LIST_ATTN as DEFAULT_TP_LIST_ATTN,
+    )
+    from common_test_cases import (
+        DSV4_FLASH_SPARSE_KERNELS as KERNELS,
+    )
+    from common_test_cases import (
+        _build_dsv4_flash_sparse_test_cases as _build_sparse_test_cases,
+    )
+
```
[the remark this answers is not in the document]
Comment (YijiaZhao) [reply collector/sglang/deepseekv4_sparse_modules.py:93]: fixed

File: src/aiconfigurator/sdk/perf_database.py [hunk 5]
```diff
@@ -1269,17 +1269,250 @@ def load_deepseek_v4_mhc_module_data(mhc_file: str):
...
+    try:
+        exact = dict_[x][y][z]
+        if isinstance(exact, dict) and "latency" in exact:
+            return exact
+    except (KeyError, TypeError):
+        pass
+    try:
+        return self._interp_3d(x, y, z, dict_, "cubic")
+    except Exception:
+        return self._interp_3d(x, y, z, dict_, "linear")
```
[the remark this answers is not in the document]
Comment (YijiaZhao) [reply src/aiconfigurator/sdk/perf_database.py]: fixed

**Review by tianhaox — APPROVED** [review]
